File: clipparse/clipvalidate.cpp
// `.clip` の参照整合性を検査する (tools/clip_validate.py の C++ 版)。
//
// `ClipFile::checkAll` はブロックの読み出しを検査するが、こちらは
// **SQLite の中の参照が全部生きているか**を見る。CSP で開く前に通すこと。
// ここに引っかかる種類の間違いは、寛容なリーダでは読めてしまうのに
// CSP では落ちたり全面透明になったりする。

#include "clipwrite.h"
#include "sqlite3.h"

#include <cstdarg>
#include <cstdio>
#include <functional>
#include <set>
#include <string>

namespace clip {

  namespace {

    std::set<int64_t> idSet(sqlite3 *db, const char *sql) {
      std::set<int64_t> out;
      sqlite3_stmt *st = nullptr;
      if (sqlite3_prepare_v2(db, sql, -1, &st, nullptr) != SQLITE_OK) return out;
      while (sqlite3_step(st) == SQLITE_ROW) out.insert(sqlite3_column_int64(st, 0));
      sqlite3_finalize(st);
      return out;
    }

    int64_t one(sqlite3 *db, const char *sql, int64_t arg = -1) {
      sqlite3_stmt *st = nullptr;
      int64_t v = 0;
      if (sqlite3_prepare_v2(db, sql, -1, &st, nullptr) != SQLITE_OK) return 0;
      if (arg >= 0) sqlite3_bind_int64(st, 1, arg);
      if (sqlite3_step(st) == SQLITE_ROW) v = sqlite3_column_int64(st, 0);
      sqlite3_finalize(st);
      return v;
    }

    std::string fmt(const char *f, ...) {
      char buf[512];
      va_list ap;
      va_start(ap, f);
      vsnprintf(buf, sizeof buf, f, ap);
      va_end(ap);
      return buf;
    }

  } // namespace

  bool validate(const char *path, std::vector<std::string> &problems) {
    problems.clear();
    ClipWriter w;
    if (!w.load(path)) {
      problems.push_back("cannot load: " + w.error());
      return false;
    }
    sqlite3 *db = w.db();

    const std::set<int64_t> layers = idSet(db, "SELECT MainId FROM Layer");
    const std::set<int64_t> mipmaps = idSet(db, "SELECT MainId FROM Mipmap");
    const std::set<int64_t> infos = idSet(db, "SELECT MainId FROM MipmapInfo");
    const std::set<int64_t> offs = idSet(db, "SELECT MainId FROM Offscreen");
    const std::set<int64_t> thumbs = idSet(db, "SELECT MainId FROM LayerThumbnail");

    const int64_t root = one(db, "SELECT CanvasRootFolder FROM Canvas");
    const int64_t cur = one(db, "SELECT CanvasCurrentLayer FROM Canvas");
    if (!layers.count(root))
      problems.push_back(fmt("Canvas.CanvasRootFolder=%lld is not in Layer",
                             (long long)root));
    if (cur && !layers.count(cur))
      problems.push_back(fmt("Canvas.CanvasCurrentLayer=%lld is not in Layer"
                             " (points at a deleted layer)", (long long)cur));

    // --- ツリー: 閉路と到達性 ---
    std::set<int64_t> seen;
    std::function<void(int64_t)> walk = [&](int64_t node) {
      while (node) {
        if (seen.count(node)) {
          problems.push_back(fmt("layer sibling chain loops at #%lld", (long long)node));
          return;
        }
        if (!layers.count(node)) {
          problems.push_back(fmt("chain points at missing layer #%lld", (long long)node));
          return;
        }
        seen.insert(node);
        const int64_t kid = one(db, "SELECT LayerFirstChildIndex FROM Layer"
                                    " WHERE MainId=?", node);
        const int64_t nxt = one(db, "SELECT LayerNextIndex FROM Layer WHERE MainId=?",
                                node);
        if (kid) walk(kid);
        node = nxt;
      }
    };
    if (layers.count(root)) {
      seen.insert(root);
      walk(one(db, "SELECT LayerFirstChildIndex FROM Layer WHERE MainId=?", root));
    }
    for (int64_t id : layers)
      if (!seen.count(id))
        problems.push_back(fmt("layer #%lld is unreachable from the tree",
                               (long long)id));

    // --- ミップ連鎖 ---
    sqlite3_stmt *st = nullptr;
    if (sqlite3_prepare_v2(db, "SELECT MainId, LayerRenderMipmap,"
                               " LayerRenderThumbnail FROM Layer",
                           -1, &st, nullptr) == SQLITE_OK) {
      while (sqlite3_step(st) == SQLITE_ROW) {
        const int64_t lid = sqlite3_column_int64(st, 0);
        const int64_t mm = sqlite3_column_int64(st, 1);
        const int64_t tn = sqlite3_column_int64(st, 2);
        if (mm && !mipmaps.count(mm)) {
          problems.push_back(fmt("Layer #%lld.LayerRenderMipmap=%lld is not in Mipmap",
                                 (long long)lid, (long long)mm));
        } else if (mm) {
          int hops = 0;
          const int64_t count = one(db, "SELECT MipmapCount FROM Mipmap WHERE MainId=?",
                                    mm);
          for (int64_t node = one(db, "SELECT BaseMipmapInfo FROM Mipmap WHERE MainId=?",
                                  mm);
               node;) {
            if (!infos.count(node)) {
              problems.push_back(fmt("Layer #%lld mipmap chain breaks at MipmapInfo"
                                     " #%lld", (long long)lid, (long long)node));
              break;
            }
            const int64_t o = one(db, "SELECT Offscreen FROM MipmapInfo WHERE MainId=?",
                                  node);
            if (!offs.count(o))
              problems.push_back(fmt("MipmapInfo #%lld.Offscreen=%lld is missing",
                                     (long long)node, (long long)o));
            node = one(db, "SELECT NextIndex FROM MipmapInfo WHERE MainId=?", node);
            if (++hops > 64) {
              problems.push_back(fmt("Layer #%lld mipmap chain does not terminate",
                                     (long long)lid));
              break;
            }
          }
          // **段数と食い違うと CSP が読み込み中に落ちる**
          if (count != hops)
            problems.push_back(fmt("Mipmap #%lld.MipmapCount=%lld != actual levels %d"
                                   " (CSP walks past the end and crashes)",
                                   (long long)mm, (long long)count, hops));
        }
        if (tn && !thumbs.count(tn)) {
          problems.push_back(fmt("Layer #%lld.LayerRenderThumbnail=%lld is missing",
                                 (long long)lid, (long long)tn));
        } else if (tn) {
          const int64_t o = one(db, "SELECT ThumbnailOffscreen FROM LayerThumbnail"
                                    " WHERE MainId=?", tn);
          if (!offs.count(o))
            problems.push_back(fmt("LayerThumbnail #%lld.ThumbnailOffscreen=%lld"
                                   " is missing", (long long)tn, (long long)o));
        }
      }
      sqlite3_finalize(st);
    }

    // --- 孤児行 ---
    const char *tables[] = { "Offscreen", "Mipmap", "MipmapInfo", "LayerThumbnail" };
    for (const char *t : tables) {
      const std::string sql = std::string("SELECT COUNT(*) FROM [") + t +
                              "] WHERE LayerId NOT IN (SELECT MainId FROM Layer)";
      const int64_t n = one(db, sql.c_str());
      if (n)
        problems.push_back(fmt("%s has %lld orphan rows (LayerId not in Layer)",
                               t, (long long)n));
    }

    // --- 格納型 ---
    // **同じ ID なのにテーブルごとに BLOB / TEXT と違う**。取り違えると
    // CSP は実体を解決できず、そのレイヤを全面透明として開く。
    {
      std::set<std::string> kinds;
      if (sqlite3_prepare_v2(db, "SELECT DISTINCT typeof(BlockData) FROM Offscreen",
                             -1, &st, nullptr) == SQLITE_OK) {
        while (sqlite3_step(st) == SQLITE_ROW)
          kinds.insert((const char *)sqlite3_column_text(st, 0));
        sqlite3_finalize(st);
      }
      if (kinds.size() != 1 || !kinds.count("blob"))
        problems.push_back("Offscreen.BlockData must be stored as blob");

      kinds.clear();
      if (sqlite3_prepare_v2(db, "SELECT DISTINCT typeof(ExternalID) FROM ExternalChunk",
                             -1, &st, nullptr) == SQLITE_OK) {
        while (sqlite3_step(st) == SQLITE_ROW)
          kinds.insert((const char *)sqlite3_column_text(st, 0));
        sqlite3_finalize(st);
      }
      if (!kinds.empty() && (kinds.size() != 1 || !kinds.count("text")))
        problems.push_back("ExternalChunk.ExternalID must be stored as text");
    }

    // --- 行数と実チャンク数 ---
    const int64_t rows = one(db, "SELECT COUNT(*) FROM ExternalChunk");
    if (rows != w.externalCount())
      problems.push_back(fmt("ExternalChunk has %lld rows but the file has %lld chunks",
                             (long long)rows, (long long)w.externalCount()));
    return problems.empty();
  }

} // namespace clip

File: tools/clip_validate.py
"""`.clip` の参照整合性を検査する。

    python tools/clip_validate.py a.clip b.clip ...

`clip_cli --check` はブロックの読み出しを検査するが、こちらは
**SQLite の中の参照が全部生きているか**を見る。CSP が落ちるファイルと
開けるファイルの差を機械的に洗い出すために書いた。

見るもの:

- `Canvas` が指すレイヤ (ルート / カレント) が実在するか
- レイヤの兄弟・子チェーンに**閉路**が無いか、全レイヤに到達できるか
- `LayerRenderMipmap` → `Mipmap` → `MipmapInfo` 連鎖 → `Offscreen` が全部生きているか
- `Mipmap.MipmapCount` が実際の段数と一致するか (**違うと CSP が落ちる**)
- `LayerThumbnail.ThumbnailOffscreen` が生きているか
- `LayerId` が実在しない**孤児行**が無いか
- `Offscreen.BlockData` の格納型 (**BLOB でないと CSP が実体を解決できない**)
- `ExternalChunk` の行と実チャンクの数が合うか
  (どこからも参照されていないチャンクは CSP のファイルにもあるので情報扱い)

依存: 標準ライブラリのみ。
"""

import sys

try:                                        # pip 版 (clipparse_tools パッケージ内)
    from .clip_write import ClipFile, as_str
except ImportError:                         # リポジトリの tools/ から直接実行
    from clip_write import ClipFile, as_str


def _ids(cur, table, col="MainId"):
    return {r[0] for r in cur.execute(f"SELECT [{col}] FROM [{table}]")}


def validate(path, verbose=True):
    c = ClipFile(path)
    cur = c.db.cursor()
    bad = []

    layers = _ids(cur, "Layer")
    mipmaps = _ids(cur, "Mipmap")
    infos = _ids(cur, "MipmapInfo")
    offs = _ids(cur, "Offscreen")
    thumbs = _ids(cur, "LayerThumbnail")

    canvas_id, root, cur_layer = cur.execute(
        "SELECT MainId, CanvasRootFolder, CanvasCurrentLayer FROM Canvas").fetchone()
    if root not in layers:
        bad.append(f"Canvas.CanvasRootFolder={root} が Layer に無い")
    if cur_layer and cur_layer not in layers:
        bad.append(f"Canvas.CanvasCurrentLayer={cur_layer} が Layer に無い"
                   f" (削除したレイヤを指したまま)")

    # --- ツリー: 閉路と到達性 ---
    seen = set()

    def walk(node, depth=0):
        while node:
            if node in seen:
                bad.append(f"レイヤの兄弟チェーンに閉路: #{node}")
                return
            if node not in layers:
                bad.append(f"存在しないレイヤ #{node} をチェーンが指している")
                return
            seen.add(node)
            kid, nxt = cur.execute(
                "SELECT LayerFirstChildIndex, LayerNextIndex FROM Layer"
                " WHERE MainId=?", (node,)).fetchone()
            if kid:
                walk(kid, depth + 1)
            node = nxt

    if root in layers:
        seen.add(root)
        kid = cur.execute("SELECT LayerFirstChildIndex FROM Layer WHERE MainId=?",
                          (root,)).fetchone()[0]
        walk(kid)
    lost = layers - seen
    if lost:
        bad.append(f"ツリーから到達できないレイヤ: {sorted(lost)}")

    # --- ミップ連鎖 ---
    for lid, mm, tn in cur.execute(
            "SELECT MainId, LayerRenderMipmap, LayerRenderThumbnail FROM Layer"):
        if mm and mm not in mipmaps:
            bad.append(f"Layer #{lid}.LayerRenderMipmap={mm} が Mipmap に無い")
        elif mm:
            node, count = cur.execute(
                "SELECT BaseMipmapInfo, MipmapCount FROM Mipmap WHERE MainId=?",
                (mm,)).fetchone()
            hops = 0
            while node:
                if node not in infos:
                    bad.append(f"Layer #{lid} のミップ連鎖が MipmapInfo #{node} で切れる")
                    break
                o, nxt = cur.execute(
                    "SELECT Offscreen, NextIndex FROM MipmapInfo WHERE MainId=?",
                    (node,)).fetchone()
                if o not in offs:
                    bad.append(f"MipmapInfo #{node}.Offscreen={o} が Offscreen に無い")
                node = nxt
                hops += 1
                if hops > 64:
                    bad.append(f"Layer #{lid} のミップ連鎖が終わらない (閉路)")
                    break
            if count != hops:
                bad.append(f"Mipmap #{mm}.MipmapCount={count} が実際の段数 {hops} と違う"
                           f" (CSP が存在しない段まで辿って落ちる)")
        if tn and tn not in thumbs:
            bad.append(f"Layer #{lid}.LayerRenderThumbnail={tn} が LayerThumbnail に無い")
        elif tn:
            o = cur.execute("SELECT ThumbnailOffscreen FROM LayerThumbnail"
                            " WHERE MainId=?", (tn,)).fetchone()[0]
            if o not in offs:
                bad.append(f"LayerThumbnail #{tn}.ThumbnailOffscreen={o} が無い")

    # --- 孤児 ---
    for table in ("Offscreen", "Mipmap", "MipmapInfo", "LayerThumbnail"):
        orphan = sorted({r[0] for r in cur.execute(
            f"SELECT MainId FROM [{table}] WHERE LayerId NOT IN"
            f" (SELECT MainId FROM Layer)")})
        if orphan:
            bad.append(f"{table} に孤児行 (LayerId が Layer に無い): {orphan[:12]}"
                       f"{' ...' if len(orphan) > 12 else ''} 計 {len(orphan)}")

    # --- 格納型 ---
    t = [r[0] for r in cur.execute("SELECT DISTINCT typeof(BlockData) FROM Offscreen")]
    if sorted(t) != ["blob"]:
        bad.append(f"Offscreen.BlockData の格納型が {t} (CSP は blob のみ)")
    t = [r[0] for r in cur.execute(
        "SELECT DISTINCT typeof(ExternalID) FROM ExternalChunk")]
    if t and sorted(t) != ["text"]:
        bad.append(f"ExternalChunk.ExternalID の格納型が {t} (CSP は text のみ)")

    # --- 実体 ---
    have = {as_str(e) for e, in cur.execute("SELECT ExternalID FROM ExternalChunk")}
    # 参照元は `Offscreen.BlockData` だけではない。`ExternalTableAndColumnName` が
    # 「この列に external_id が入る」を宣言しているので、そこを全部見る
    # (実運用ファイルでは Track / TimeLapseBlob / Canvas3DModelBank などが持つ)。
    used = set()
    for table, col in cur.execute(
            "SELECT TableName, ColumnName FROM ExternalTableAndColumnName"):
        try:
            used |= {as_str(b) for b, in cur.execute(
                f"SELECT [{col}] FROM [{table}] WHERE [{col}] IS NOT NULL")}
        except Exception:
            bad.append(f"ExternalTableAndColumnName の {table}.{col} が引けない")
    # **未参照のチャンクは異常ではない**。CSP 自身のファイルにも残っている
    # (haruse-ja に 2 件、nazoani01 に 52 件。`RemovedExternal` は空)。
    # 編集で捨てられた実体を回収していないだけなので、情報として出すに留める。
    note = (f"どの列からも指されていない ExternalChunk が {len(have - used)} 件 "
            f"(CSP のファイルにもあるので異常ではない)") if have - used else None
    chunk_ids = {as_str(e) for e, _p in c.externals}
    if chunk_ids != have:
        bad.append(f"ExternalChunk の行 {len(have)} 件と実チャンク {len(chunk_ids)} 件が食い違う")

    if verbose:
        print(f"{path}")
        print(f"  レイヤ {len(layers)} / Offscreen {len(offs)} / 実体 {len(chunk_ids)}")
        if note:
            print(f"  -- {note}")
        for b in bad:
            print(f"  NG {b}")
        if not bad:
            print("  => 参照整合性 OK")
    c.close()
    return bad


def main(argv=None):
    argv = argv or sys.argv[1:]
    if not argv:
        print(__doc__)
        return 2
    return 1 if [p for p in argv if validate(p)] else 0


if __name__ == "__main__":
    sys.path.insert(0, __file__.rsplit("\\", 1)[0].rsplit("/", 1)[0])
    sys.exit(main())
